File: typescript/packages/core/src/core/airtable/normalize.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { jsonBytes, jsonlBytes } from '../render/json.ts'

export type Row = Record<string, unknown>

/** Whether a decoded JSON value is an object. */
export function isRow(value: unknown): value is Row {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

/** A decoded JSON object, or an empty one for anything else. */
export function asRow(value: unknown): Row {
  return isRow(value) ? value : {}
}

/** The objects of a decoded JSON array; anything else holds none. */
export function asRows(value: unknown): Row[] {
  return Array.isArray(value) ? value.filter(isRow) : []
}

function field(row: Row, key: string): unknown {
  return row[key] ?? null
}

/** One base as the listing names it: id, name and permission level. */
export function normalizeBaseSummary(base: Row): Row {
  return {
    base_id: field(base, 'id'),
    base_name: field(base, 'name'),
    permission_level: field(base, 'permissionLevel'),
  }
}

/** base.json: the base and the tables it holds. */
export function normalizeBase(base: Row, tables: readonly Row[]): Row {
  return {
    ...normalizeBaseSummary(base),
    tables: tables.map((table) => ({
      table_id: field(table, 'id'),
      table_name: field(table, 'name'),
      primary_field_id: field(table, 'primaryFieldId'),
    })),
  }
}

/** One column: its id, name, type and the type's options. */
export function normalizeField(column: Row): Row {
  return {
    field_id: field(column, 'id'),
    field_name: field(column, 'name'),
    type: field(column, 'type'),
    description: field(column, 'description'),
    options: field(column, 'options'),
  }
}

/** One saved view: its id, name and kind (grid, kanban, ...). */
export function normalizeView(view: Row): Row {
  return {
    view_id: field(view, 'id'),
    view_name: field(view, 'name'),
    type: field(view, 'type'),
  }
}

/** table.json: the column schema and the saved views. */
export function normalizeTable(table: Row, baseId: string): Row {
  return {
    table_id: field(table, 'id'),
    table_name: field(table, 'name'),
    base_id: baseId,
    description: field(table, 'description'),
    primary_field_id: field(table, 'primaryFieldId'),
    fields: asRows(table.fields).map(normalizeField),
    views: asRows(table.views).map(normalizeView),
  }
}

/**
 * One records.jsonl row. `fields` is Airtable's cell map, keyed by field
 * name and passed through untouched: an empty cell is absent (Airtable never
 * sends one), a link is a list of record ids, and an attachment's url is the
 * API's own, which expires two hours after it was fetched.
 */
export function normalizeRecord(record: Row): Row {
  return {
    record_id: field(record, 'id'),
    created_time: field(record, 'createdTime'),
    fields: asRow(record.fields),
  }
}

/** One comment on a record, with its author flattened. */
export function normalizeComment(comment: Row): Row {
  const who = asRow(comment.author)
  return {
    comment_id: field(comment, 'id'),
    author_id: field(who, 'id'),
    author_email: field(who, 'email'),
    author_name: field(who, 'name'),
    text: field(comment, 'text'),
    created_time: field(comment, 'createdTime'),
    last_updated_time: field(comment, 'lastUpdatedTime'),
  }
}

/** One deleted record, as the delete answer names it. */
export function normalizeDeletion(row: Row): Row {
  return { record_id: field(row, 'id'), deleted: field(row, 'deleted') }
}

/** Render a .json leaf. */
export function toJsonBytes(value: unknown): Uint8Array {
  return jsonBytes(value)
}

/** Render records one per line, in the order they were listed. */
export function recordsJsonl(listed: readonly Row[]): Uint8Array {
  return jsonlBytes(listed.map(normalizeRecord))
}

/** Render deleted records one per line, in the order they were deleted. */
export function deletionsJsonl(rows: readonly Row[]): Uint8Array {
  return jsonlBytes(rows.map(normalizeDeletion))
}

File: python/mirage/core/airtable/normalize.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.core.render.json import json_bytes, jsonl_bytes


def normalize_base_summary(base: dict[str, Any]) -> dict[str, Any]:
    """One base as the listing names it: id, name and permission level.

    Args:
        base (dict[str, Any]): the base-listing row.
    """
    return {
        "base_id": base.get("id"),
        "base_name": base.get("name"),
        "permission_level": base.get("permissionLevel"),
    }


def normalize_base(base: dict[str, Any],
                   tables: list[dict[str, Any]]) -> dict[str, Any]:
    """base.json: the base and the tables it holds.

    Args:
        base (dict[str, Any]): the base-listing row.
        tables (list[dict[str, Any]]): the base's schema tables.
    """
    return {
        **normalize_base_summary(base),
        "tables": [{
            "table_id": table.get("id"),
            "table_name": table.get("name"),
            "primary_field_id": table.get("primaryFieldId"),
        } for table in tables],
    }


def normalize_field(field: dict[str, Any]) -> dict[str, Any]:
    """One column: its id, name, type and the type's options.

    Args:
        field (dict[str, Any]): a schema field.
    """
    return {
        "field_id": field.get("id"),
        "field_name": field.get("name"),
        "type": field.get("type"),
        "description": field.get("description"),
        "options": field.get("options"),
    }


def normalize_view(view: dict[str, Any]) -> dict[str, Any]:
    """One saved view: its id, name and kind (grid, kanban, ...).

    Args:
        view (dict[str, Any]): a schema view.
    """
    return {
        "view_id": view.get("id"),
        "view_name": view.get("name"),
        "type": view.get("type"),
    }


def normalize_table(table: dict[str, Any], base_id: str) -> dict[str, Any]:
    """table.json: the column schema and the saved views.

    Args:
        table (dict[str, Any]): a schema table.
        base_id (str): the base holding it.
    """
    return {
        "table_id": table.get("id"),
        "table_name": table.get("name"),
        "base_id": base_id,
        "description": table.get("description"),
        "primary_field_id": table.get("primaryFieldId"),
        "fields": [normalize_field(f) for f in table.get("fields") or []],
        "views": [normalize_view(v) for v in table.get("views") or []],
    }


def normalize_record(record: dict[str, Any]) -> dict[str, Any]:
    """One records.jsonl row.

    ``fields`` is Airtable's cell map, keyed by field name and passed
    through untouched: an empty cell is absent (Airtable never sends
    one), a link is a list of record ids, and an attachment's url is
    the API's own, which expires two hours after it was fetched.

    Args:
        record (dict[str, Any]): a listed record.
    """
    return {
        "record_id": record.get("id"),
        "created_time": record.get("createdTime"),
        "fields": record.get("fields") or {},
    }


def normalize_comment(comment: dict[str, Any]) -> dict[str, Any]:
    """One comment on a record, with its author flattened.

    Args:
        comment (dict[str, Any]): a listed or created comment.
    """
    author = comment.get("author")
    who = author if isinstance(author, dict) else {}
    return {
        "comment_id": comment.get("id"),
        "author_id": who.get("id"),
        "author_email": who.get("email"),
        "author_name": who.get("name"),
        "text": comment.get("text"),
        "created_time": comment.get("createdTime"),
        "last_updated_time": comment.get("lastUpdatedTime"),
    }


def normalize_deletion(row: dict[str, Any]) -> dict[str, Any]:
    """One deleted record, as the delete answer names it.

    Args:
        row (dict[str, Any]): an ``{id, deleted}`` answer row.
    """
    return {"record_id": row.get("id"), "deleted": row.get("deleted")}


def to_json_bytes(value: Any) -> bytes:
    """Render a .json leaf.

    Args:
        value (Any): the normalized payload.
    """
    return json_bytes(value)


def records_jsonl(records: list[dict[str, Any]]) -> bytes:
    """Render records one per line, in the order they were listed.

    Args:
        records (list[dict[str, Any]]): listed records.
    """
    return jsonl_bytes([normalize_record(r) for r in records])


def deletions_jsonl(rows: list[dict[str, Any]]) -> bytes:
    """Render deleted records one per line, in the order they were deleted.

    Args:
        rows (list[dict[str, Any]]): delete answer rows.
    """
    return jsonl_bytes([normalize_deletion(r) for r in rows])
